apiserver/facades/client/modelgeneration/modelgeneration.go

```diff
@@ -148,19 +150,54 @@ func (api *API) CommitBranch(arg params.BranchArg) (params.IntResult, error) {
 		return result, common.ErrPerm
 	}
 
-	generation, err := api.model.Branch(arg.BranchName)
+	branch, err := api.model.Branch(arg.BranchName)
 	if err != nil {
-		return result, errors.Trace(err)
+		return intResultsError(err)
+	}
+
+	if err = api.applyBranchConfig(branch); err != nil {
+		return intResultsError(err)
 	}
 
-	if genId, err := generation.Commit(api.apiUser.Name()); err != nil {
+	if genId, err := branch.Commit(api.apiUser.Name()); err != nil {
 		result.Error = common.ServerError(err)
 	} else {
 		result.Result = genId
 	}
 	return result, nil
 }
 
+// applyBranchConfig applies configuration deltas from a branch to the
+// appropriate application charm settings documents and writes them back
+// to the DB. This writes the ultimate data as a single transaction,
+// so a branch is either committed without error, or there are no changes.
+func (api *API) applyBranchConfig(branch Generation) error {
```

Comment:
This logic is in the wrong place.

The application of the branch config to the various applications needs to be part of the `branch.Commit` work. Otherwise we are splitting the transaction.

Reply:
Yep, done.